### ISSUE TITLE
many unwanted files are bundled

### ISSUE THREAD

**Issue by farfromrefug** [NativeScript/NativeScript#9113]
That [line](https://github.com/NativeScript/NativeScript/blob/master/packages/webpack/bundle-config-loader.ts#L7) creates very big bundles with unwanted files bundled. Also it slows down the build with unwanted loaders:
* if you have css/xml as assets of your projects they will get loaded/parsed because of this. Though those are only assets!
* all css/scss are added. You might end up with duplicated files in scss case. `app.scss` includes `variables.scss`. Then `variables.scss` is already in `app.scss` but also added to the bundle because of that rule.

I thinK:
* scss / css can be removed
* flavors dont need xml
* can it be removed completely?  should it suffice to load the app entry?

**Comment by NathanaelA** [comment]
I believe that line is still valid for the default builds...

1. You do need CSS for several flavors (and all flavors can actually use them).
2. The reason we do include all CSS/SCSS is because you can dynamically load css at runtime, so all css/scss has to be bundled because we have no idea if the app will do anything dynamically with them at any point.
3.  And xml is used in several plugins because XML can also be used by all flavors, I know I've used it in a couple plugins to generate a screen or dialog, and I've seen other plugins that used it for the same purpose.

**Comment by farfromrefug** [comment]
@NathanaelA 
1/ actually you dont. removed it and my app.scss/css is still loaded perfectly. 
2/ they would get included as they would be referenced if dynamically loaded!
3/ same thing.

Webpack is smart and would include/load/parse everything you require. I already locally removed css/xml/scss from the rule and it works perfectly.

**Comment by NathanaelA** [comment]
1. Not 100% sure, but this was the case not that long ago -- if the css is not being webpacked in, then it won't be loaded.   Anything that isn't specifically referenced in its exact form won't be included.
2. Webpack is NOT smart enough for dynamic loads.  i.e.  `const theme = loadUserThemeChoice({default: "blah"}); Theme.set(theme);` would load blah.css dynamically if user hasn't picked a choice.   This is used in several apps that I'm aware of since the ns-theme plugin supports loading themes dynamically at runtime from both raw css and webpacked css.
3. Depends on the XML, if it is again dynamically loaded like the css example above it won't be included.   

The purpose of the above line is to handle anything dynamic that webpack very frequently fails at figuring out...   However, not sure about scss; scss used to be converted to css so that only css was included in the final app..

**Comment by farfromrefug** [comment]
@NathanaelA i understand what you say about XML ans thèmes loading. Still i think that rule is dead wrong as it loads everything un your app folder! 
Maybe we
 should rethink theme loading or XML requiring so that webpack can pick it up and so that we can remove that rule !

**Comment by NathanaelA** [comment]
To be honest, I'm not really understanding this issue I'm not sure why it would be any larger in most cases; any XML or CSS in your actual app/ folder (or its sub-folders) should already normally be part of your app.   So they would be found via normal webpack means (unless they are dynamic).     What XML/CSS files are in your app/** folder that aren't actually part of your app?   Shouldn't you be deleting them and keeping your app folders clean if they aren't actually part of your app?

**Comment by farfromrefug** [comment]
If you have apps.scss which includes variables.scss then it will be there twice in your bundle!
I had friends who reported this too. Now I know why. The sass loader will insert the variable.scss directly in your app.scss while this rule will add it again in your bundle.
Same thing if you have assets CSS used for example for local HTML through webview.
Also in case you have XML for another purpose as assets (like I do) you will get tons of loader errors because of that rule. Especially with flavors where there is no XML loader.
I think almost everyone is getting bigger bundle or slower build because of that rule without knowing it.

**Comment by NathanaelA** [comment]
Since I don't use Sass, never seen it myself -- but then my guess is that the scss part of that rule probably should be pulled out.   
Just the xml/css should be left.

**Comment by farfromrefug** [comment]
@NathanaelA no it can't be pulled out. 
@rigor789 did you see that working on webpack5?

**Comment by NathanaelA** [comment]
> If you have apps.scss which includes variables.scss then it will be there twice in your bundle!

No it won't the app.scss will include variables.scss and it should be parsed by the sass into its pieces and applied to app.scss so things like `$blue` becomes `#0000ff` directly -- to my knowledge sass is smart enough NOT to just embed all of variables into the app.scss.   So it isn't actually App.scss+variables.scss, it is app.scss with values replaced from variables.scss.    However, when you the above rule hits and does parse variables.scss by itself; you should get a really tiny file left over because all the variable definitions should be parsed out and so the only thing left should be basically a empty file.   Technically it isn't needed in the webpack output and maybe we should specifically exclude "variables.scss", but really it should be so tiny it is inconsequential to the added build overhead of adding additional parsing to ignore variables.scss.     However the base reason this rule exists is the same reason why the css rule does.   Webpack has no idea what scss will be used and what scss is "extra".   


> Also in case you have XML for another purpose as assets (like I do) you will get tons of loader errors because of that rule. Especially with flavors where there is no XML loader.

Then just rename your .XML `AwesomeBlah.xml` files to something else like `AwesomeBlah.assets` and be done with it.   :grinning:       No need to try and complicate the webpack since all the flavors can use XML files for building screens.

**Comment by farfromrefug** [comment]
@NathanaelA sorry Nathanel but really your answers dont help a bit. Of course if I say all that it is because I can't rename my XML... 
And actually the webpack is complicated as it is ! This is the point of this issue. I also figured out today that even if I have dev ts file which I don't intend to use yet (not imported anywhere) they are also bundle because of that rule. I am facing this in common folders shared by multiple apps. 
Having a sync recursive rule in webpack is mainly always a bad thing. They should be avoided as much as possible. 
And the last thing I would say is that the user has absolutely no control over that rule because it Is in the bundle loader. Another reason why we should rewrite it differently.

**Comment by NathanaelA** [comment]
I don't disagree that all the rules should be in the main webpack file.  I've ran into issues where the bundler didn't include files I needed, had to create a custom bundler plugin.  So I do understand the pain, and I hated doing it -- but when doing non-standard things -- it is on me to bear the pain.    However, part of the webpack 5 rewrite is to handle this and other non-standard scenarios.    

But in the current version this rule needs to stay otherwise we break a large number of applications that do depend on it.   So far the complaints you have raised have been minimal or invalid.

For example, I actually do have a app that I built that uses scss (didn't recall doing it, but after a directory scan found a couple of them).    I scanned its built files and not a single added copy of "_app-variables.scss" or "_app-common.scss" in the bundle or vendor js files.    However both these files were used in the app as `@import` support files.

![image](https://user-images.githubusercontent.com/850871/103176176-28ce3000-4835-11eb-99f5-6d8e61356878.png)

So no duplicate files....    Meaning in **normal** and the **majority** of the application usage the rules work exactly as intended and do exactly what we want and need them to do, **WITHOUT** any **duplication**. 

As for the other issue,  we have no control over where users will put their .XML or .CSS files; so we by default MUST include them all as they all are probably valid in the vast vast vast majority of NS application. So we really can't break the majority of the NativeScript eco-system because you decided to name some asset files with a .xml extension and are unwilling to give them a different extension.   My suggestion is modify your own bundler and add exceptions to exclude the files you don't want bundled until the WP 5 rewrite it released...

I'm closing this as invalid as:
1. Part of this is a **self-inflicted**, and could be easily fixed on your side if you really wanted to, and would break the vanilla eco-system if we eliminated the default behavior.  
2. Part of this appears to be invalid as I couldn't see duplicated resources in any of the apps I had that used scss. 
3. Finally Webpack 5 rewrite should be handling this and allow you to easily override the default rules when you decide to do something that it **non-standard** in a NS application.
